Use a vector backed set for speed.

baldersheim opened the pull request:
@havardpe PR
Superceeds #14042

baldersheim committed:
Use a vector backed set for speed.

havardpe commented on document/src/vespa/document/base/field.h:
```diff
@@ -28,16 +28,35 @@ class Field final : public vespalib::FieldBase,
+        class Builder {
+        public:
+            Builder & reserve(size_t sz) { _vector.reserve(sz); return *this; }
+            Builder & insert(CPtr field) { _vector.emplace_back(field); return *this; }
```
I would prefer 'add' over 'insert' here, since it implies less. Also, I would prefer 'push_back' over 'emplace_back' for values that already have the appropriate type.

baldersheim replied:
Fixed

havardpe commented on document/src/vespa/document/base/field.cpp:
```diff
@@ -7,9 +7,21 @@
+Field::Set::Set(std::vector<CPtr> fields)
+    : _fields(std::move(fields))
+{
+    std::sort(_fields.begin(), _fields.end(), Field::FieldPtrComparator());
```
consider doing unique as well, to ensure set property

baldersheim replied:
Yes, thanks.

havardpe commented on document/src/vespa/document/base/field.h:
```diff
@@ -28,16 +28,35 @@ class Field final : public vespalib::FieldBase,
+        private:
+            std::vector<CPtr> _vector;
+        };
+        bool contains(const Field & field) const;
```
consider adding a contains taking a Field::Set as well since it can be implemented more efficiently using the std::includes function (compared to calling contains per field in the other set)

baldersheim replied:
Good idea, done.

havardpe commented on document/src/vespa/document/fieldset/fieldsets.cpp:
```diff
@@ -38,25 +39,10 @@ FieldCollection::contains(const FieldSet& fields) const
-            }
-
-            return true;
+            return std::includes(_set.begin(), _set.end(), coll.getFields().begin(), coll.getFields().end(), Field::FieldPtrComparator());
```
this should be a function on the Field::Set (see other comment)

baldersheim replied:
Done

havardpe commented on searchcore/src/vespa/searchcore/proton/server/documentretriever.cpp:
```diff
@@ -82,14 +82,15 @@ ::DocumentRetriever(const DocTypeName &docTypeName,
       _doc_store(doc_store),
       _possiblePositionFields(),
-      _attributeFields(),
+      _attributeFields(document::Field::Set::Builder().build()),
```
consider adding a static convenience to make this case simpler.

baldersheim replied:
emptySet

havardpe reviewed (commented):
looks good. some minor comments

baldersheim committed:
Make sure the entries in the set are unique.
Make both a less and equal operator so std::unique can be used to.
Add and interface so that whole sets can be tested effectively if they are subsets.
